### PR TITLE
[REVIEW] Device memory size for string category

### PR DESCRIPTION
**Pull request by nartal1**
This PR is to calculate the device memory size for string category columns. This does approximate calculation by assumption that the string length in each row is 10.

### CODE REVIEW

File: java/src/main/java/ai/rapids/cudf/ColumnVector.java [hunk 1]
```diff
@@ -1863,12 +1863,18 @@ public void setHostData(BufferEncapsulator<HostMemoryBuffer> hostData) {
       return deviceData;
     }
 
+    /**
+     * This returns total memory allocated in device for the ColumnVector.
+     * If DTYpe is STRING_CATEGORY, approximate size is returned by guessing. It assumes
+     * length of strings to be 10 in each row and returns 240 bytes per row.
+     * @param type
+     * @return memoru allocated in device
```
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/ColumnVector.java]:
```suggestion
     * @return number of device bytes allocated for this column
```

File: java/src/main/java/ai/rapids/cudf/ColumnVector.java [hunk 2]
```diff
@@ -1863,12 +1863,18 @@ public void setHostData(BufferEncapsulator<HostMemoryBuffer> hostData) {
       return deviceData;
     }
 
+    /**
+     * This returns total memory allocated in device for the ColumnVector.
+     * If DTYpe is STRING_CATEGORY, approximate size is returned by guessing. It assumes
```
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/ColumnVector.java]:
```suggestion
     * NOTE: If DType is STRING_CATEGORY, the size is estimated. The estimate assumes the
```

File: java/src/main/java/ai/rapids/cudf/ColumnVector.java [hunk 3]
```diff
@@ -1863,12 +1863,18 @@ public void setHostData(BufferEncapsulator<HostMemoryBuffer> hostData) {
...
+     * length of strings to be 10 in each row and returns 240 bytes per row.
```
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/ColumnVector.java]:
```suggestion
     * length of strings to be 10 characters in each row.
```

File: java/src/main/native/src/ColumnVectorJni.cpp [hunk 4]
```diff
@@ -573,8 +573,14 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_getDeviceMemoryStringSi
     if (dtype == GDF_STRING) {
       NVStrings *nvstr = static_cast<NVStrings *>(column->data);
       return static_cast<jlong>(nvstr->memsize());
+    } else if (dtype == GDF_STRING_CATEGORY) {
+      NVCategory *cats = static_cast<NVCategory *>(column->dtype_info.category);
+      unsigned long dict_size = cats->keys_size();
+      unsigned long dict_size_total = dict_size * GDF_INT32;
+      unsigned long category_size_total = dict_size * 240;  // assumption:240B per row (10 chars in each string)
```
Comment (jlowe) [inline java/src/main/native/src/ColumnVectorJni.cpp]: It isn't clear to me how 10 characters becomes 240 bytes.  Normally 10 characters would be 10 bytes, so where are the other 230 bytes coming from?  A comment here explaining how 10 chars == 240 bytes would help.

Comment (nartal1) [reply java/src/main/native/src/ColumnVectorJni.cpp]: I think we could approximate it to 24 bytes. It's 8B aligned and considering overhead for custrings_view(8B), approximating it to 24 bytes per row.

File: java/src/main/native/src/ColumnVectorJni.cpp [hunk 5]
```diff
@@ -573,8 +573,16 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_getDeviceMemoryStringSi
     if (dtype == GDF_STRING) {
       NVStrings *nvstr = static_cast<NVStrings *>(column->data);
       return static_cast<jlong>(nvstr->memsize());
+    } else if (dtype == GDF_STRING_CATEGORY) {
+      NVCategory *cats = static_cast<NVCategory *>(column->dtype_info.category);
+      unsigned long dict_size = cats->keys_size();
+      unsigned long dict_size_total = dict_size * GDF_INT32;
+      /*NOTE: Assumption being made that strings in each row is of 10 chars. So the result would be approximate.
```
Comment (jlowe) [inline java/src/main/native/src/ColumnVectorJni.cpp]: style nit: Either use a comment style like:
```c++
// NOTE: Assumption ..l
// ...
```
or like this:
```
/* NOTE Assumption ...
 * ...
 */
```

File: java/src/main/java/ai/rapids/cudf/ColumnVector.java [hunk 6]
```diff
@@ -1863,12 +1863,18 @@ public void setHostData(BufferEncapsulator<HostMemoryBuffer> hostData) {
       return deviceData;
     }
 
+    /**
+     * This returns total memory allocated in device for the ColumnVector.
+     * NOTE: If DType is STRING_CATEGORY, the size is estimated. The estimate assumes the length
+     * of strings to be 10 characters in each row and returns 24 bytes per row.
```
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/ColumnVector.java]: Technically it's not 24 bytes per row but 24 bytes per dictionary entry.  Callers may be surprised that it's not just `row_count * 24`.

**Comment by nartal1** [comment]
@jlowe Please take another look. Thanks!

File: CHANGELOG.md [hunk 7]
```diff
@@ -48,7 +48,7 @@
 - PR #3138 Movey unary files to legacy
 - PR #3175 Set cmake cuda version variables
 - PR #3171 Move deprecated error macros to legacy
-
```
Comment (jlowe) [inline CHANGELOG.md]: This should just be an addition of a line, not also a removal of a whitespace line.

Comment (nartal1) [reply CHANGELOG.md]: Done.  PTAL.

**Pull request merged**